src/symbols_linux.cpp

```diff
@@ -845,4 +836,48 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
     }
 }
 
+static bool isValidHandle(const CodeCache* cc, void* handle) {
+    Dl_info dl_info;
+    struct link_map* map;
+
+    // Check shared objects are still available at original base image
+    if (handle == NULL || dlinfo(handle, RTLD_DI_LINKMAP, &map) != 0 || dladdr(map->l_ld, &dl_info) == 0) {
+        return false;
+    }
+
+    return cc->imageBase() == (const char*)dl_info.dli_fbase;
+}
+
+UnloadProtection::UnloadProtection(const CodeCache *cc) {
+    _lib_handle = NULL;
+    _valid = false;
+
+    if (OS::isMusl() || isMainExecutable(cc->imageBase(), cc->maxAddress()) || isLoader(cc->imageBase())) {
+        _valid = true;
+        return;
+    }
+
+    const char* stripped_name;
+    size_t name_len = strlen(cc->name());
+    if (name_len > 10 && strcmp(cc->name() + name_len - 10, " (deleted)") == 0) {
+        char* buf = (char*) alloca(name_len);
+        memcpy(buf, cc->name(), name_len);
+        buf[name_len - 10] = 0;
+        stripped_name = buf;
+    } else {
+        stripped_name = cc->name();
+    }
+
+    // Protect library from unloading while parsing in-memory ELF program headers.
+    // Also, dlopen() ensures the library is fully loaded.
+    _lib_handle = dlopen(stripped_name, RTLD_LAZY | RTLD_NOLOAD);
+    _valid = isValidHandle(cc, _lib_handle);
```

Comment:
Not the best name, IMO.
Any non-null handle is valid, but what we really check here is whether it is the same shared library that was seen at CodeCache creation time.

Reply:
- `checkHandleConsistency`
- `isHandleConsistent`

I'll think about it, let me know if you have any preference

Reply:
41d89d1a4c6b4a80b4f2a2591d051b39eea72a84